src/proxy.rs

```diff
@@ -469,6 +477,102 @@ impl Dst for Uri {
+        #[cfg(target_os = "windows")]
+        return get_from_registry();
+    }
+    return proxies;
+}
+
+fn insert_proxy(proxies: &mut HashMap<String, Url>, schema: String, addr: String)
+{
+    if let Ok(valid_addr) = Url::parse(&addr) {
+        proxies.insert(schema, valid_addr);
+    }
+}
+
+fn get_from_environment() -> HashMap<String, Url> {
+    let mut proxies: HashMap<String, Url> = HashMap::new();
+
+    const PROXY_KEY_ENDS: &str = "_proxy";
+
+    for (key, value) in env::vars() {
+        let key: String = key.to_lowercase();
+        if key.ends_with(PROXY_KEY_ENDS) {
+            let end_indx = key.len() - PROXY_KEY_ENDS.len();
+            let schema = &key[..end_indx];
+            insert_proxy(&mut proxies, String::from(schema), String::from(value));
+        }
+    }
+    return proxies;
+}
+
+
+#[cfg(target_os = "windows")]
+fn get_from_registry_impl() -> Result<HashMap<String, Url>, Box<dyn Error>> {
+    let hkcu = RegKey::predef(HKEY_CURRENT_USER);
+    let internet_setting: RegKey =
+        hkcu.open_subkey("Software\\Microsoft\\Windows\\CurrentVersion\\Internet Settings")?;
+    // ensure the proxy is enable, if the value doesn't exist, an error will returned.
+    let proxy_enable: u32 = internet_setting.get_value("ProxyEnable")?;
+    let proxy_server: String = internet_setting.get_value("ProxyServer")?;
+
+    if proxy_enable == 0 {
+        return Ok(HashMap::new());
+    }
+
+    let mut proxies: HashMap<String, Url> = HashMap::new();
+    if proxy_server.contains("=") {
+        // per-protocol settings.
+        for p in proxy_server.split(";") {
+            let protocol_parts: Vec<&str> = p.split("=").collect();
+            match protocol_parts.as_slice() {
+                [protocol, address] => {
+                    insert_proxy(&mut proxies, String::from(*protocol), String::from(*address));
+                }
+                _ => {
+                    // Contains invalid protocol setting, just break the loop
+                    // And make proxies to be empty.
+                    proxies.clear();
+                    break;
+                }
+            }
+        }
+    } else {
+        // Use one setting for all protocols.
+        if proxy_server.starts_with("http:") {
+            insert_proxy(&mut proxies, String::from("http"), proxy_server);
+        } else {
+            insert_proxy(&mut proxies, String::from("http"), format!("http://{}", proxy_server));
+            insert_proxy(&mut proxies, String::from("https"), format!("https://{}", proxy_server));
+            insert_proxy(&mut proxies, String::from("ftp"), format!("https://{}", proxy_server));
+        }
+    }
+    return Ok(proxies);
+}
+
+#[cfg(target_os = "windows")]
+fn get_from_registry() -> HashMap<String, Url> {
+    let results = get_from_registry_impl();
+    match results {
+        Ok(proxies) => proxies,
+        Err(_) => HashMap::new(),
+    }
```

Comment:
```suggestion
    results.unwrap_or(HashMap::new())
```
instead of the match would be more rust style